<a href="https://colab.research.google.com/github/annisaeka123/2341720131_ML_2025/blob/main/JS09/TP_JS09_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

TUGAS 2

Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

1. Menggunakan data spam.csv

2. Fitur CountVectorizer dengan mengaktifkan stop_words

3. Evaluasi hasilnya

Buatlah model klasfikasi Multinomial Naive Bayes dengan ketentuan,

1. Menggunakan data spam.csv

2. Fitur TF-IDF dengan mengaktifkan stop_words

3. Evaluasi hasilnya dan bandingkan dengan hasil pada Tugas no 2.

4. Berikan kesimpulan fitur mana yang terbaik pada kasus data spam.csv

Import Library dan Load Dataset

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report

# Load dataset
df = pd.read_csv('spam.csv', encoding='latin-1')

# Hapus kolom tidak relevan
df = df.drop(df.iloc[:, 2:], axis=1)
df = df.rename(columns={'v1': 'label', 'v2': 'message'})

# Encode label: spam=1, ham=0
df['label'] = df['label'].str.strip().str.lower().map({'spam': 1, 'ham': 0})

# Split data
X_train, X_test, y_train, y_test = train_test_split(df['message'], df['label'], test_size=0.2, random_state=42)


MultinomialNB dengan CountVectorizer (stop_words='english')

In [ ]:
# CountVectorizer
cv = CountVectorizer(stop_words='english')
X_train_cv = cv.fit_transform(X_train)
X_test_cv = cv.transform(X_test)

# Model Naive Bayes
mnb_cv = MultinomialNB()
mnb_cv.fit(X_train_cv, y_train)

# Prediksi dan evaluasi
y_pred_cv = mnb_cv.predict(X_test_cv)
acc_cv = accuracy_score(y_test, y_pred_cv)

print("=== MultinomialNB + CountVectorizer ===")
print(f"Akurasi: {acc_cv:.4f}")
print(classification_report(y_test, y_pred_cv))


=== MultinomialNB + CountVectorizer ===
Akurasi: 0.9839
              precision    recall  f1-score   support

           0       0.99      0.99      0.99       965
           1       0.96      0.92      0.94       150

    accuracy                           0.98      1115
   macro avg       0.97      0.96      0.96      1115
weighted avg       0.98      0.98      0.98      1115



MultinomialNB dengan TF-IDF (stop_words='english')

In [ ]:
# TF-IDF Vectorizer
tfidf = TfidfVectorizer(stop_words='english')
X_train_tfidf = tfidf.fit_transform(X_train)
X_test_tfidf = tfidf.transform(X_test)

# Model Naive Bayes
mnb_tfidf = MultinomialNB()
mnb_tfidf.fit(X_train_tfidf, y_train)

# Prediksi dan evaluasi
y_pred_tfidf = mnb_tfidf.predict(X_test_tfidf)
acc_tfidf = accuracy_score(y_test, y_pred_tfidf)

print("\n=== MultinomialNB + TF-IDF ===")
print(f"Akurasi: {acc_tfidf:.4f}")
print(classification_report(y_test, y_pred_tfidf))



=== MultinomialNB + TF-IDF ===
Akurasi: 0.9668
              precision    recall  f1-score   support

           0       0.96      1.00      0.98       965
           1       1.00      0.75      0.86       150

    accuracy                           0.97      1115
   macro avg       0.98      0.88      0.92      1115
weighted avg       0.97      0.97      0.96      1115



Evaluasi dan Perbandingan Hasil

Berdasarkan hasil pengujian, model Multinomial Naive Bayes dengan CountVectorizer menghasilkan akurasi sebesar 98.39%, sedangkan model dengan TF-IDF menghasilkan akurasi sedikit lebih rendah yaitu 96.68%. Model dengan CountVectorizer juga menunjukkan nilai recall dan f1-score yang lebih seimbang antara kelas spam dan ham, dengan f1-score masing-masing sebesar 0.99 untuk kelas ham dan 0.94 untuk kelas spam. Sementara itu, model TF-IDF menunjukkan performa sangat baik pada kelas ham (recall 1.00) namun relatif lebih rendah pada kelas spam (recall 0.75), yang berarti beberapa pesan spam masih salah diklasifikasikan sebagai ham.

Perbedaan performa ini terjadi karena CountVectorizer hanya menghitung frekuensi kemunculan kata tanpa mempertimbangkan bobot relatif antar dokumen, sehingga cocok untuk dataset spam.csv yang sederhana dan memiliki pola kata spam yang sering berulang. Sebaliknya, TF-IDF memberikan bobot lebih tinggi pada kata-kata unik dan menurunkan bobot kata umum, sehingga lebih cocok untuk dataset teks yang lebih kompleks dan beragam. Dalam kasus ini, karena banyak kata spam muncul berulang dengan pola serupa, CountVectorizer dapat mengenali pola tersebut dengan lebih baik sehingga memberikan akurasi yang lebih tinggi.

Kesimpulan

Dari hasil evaluasi dan perbandingan, dapat disimpulkan bahwa fitur terbaik untuk kasus klasifikasi pesan spam pada dataset spam.csv adalah CountVectorizer dengan stop_words aktif. Fitur ini memberikan performa yang lebih stabil dan akurasi lebih tinggi dibandingkan TF-IDF pada dataset dengan pola kata yang berulang dan sederhana. Namun, untuk dataset teks yang lebih kompleks atau dengan variasi kata yang tinggi, TF-IDF tetap menjadi pilihan yang lebih unggul karena mampu menonjolkan kata-kata yang paling relevan terhadap kelas tertentu.